com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/LambdaExpressionLocator.java

```diff
@@ -0,0 +1,109 @@
+/*******************************************************************************
+* Copyright (c) 2022 Microsoft Corporation and others.
+* All rights reserved. This program and the accompanying materials
+* are made available under the terms of the Eclipse Public License v1.0
+* which accompanies this distribution, and is available at
+* http://www.eclipse.org/legal/epl-v10.html
+*
+* Contributors:
+*     Gayan Perera (gayanper@gmail.com) - initial API and implementation
+*******************************************************************************/
+package com.microsoft.java.debug;
+
+import org.eclipse.jdt.core.dom.ASTNode;
+import org.eclipse.jdt.core.dom.ASTVisitor;
+import org.eclipse.jdt.core.dom.AbstractTypeDeclaration;
+import org.eclipse.jdt.core.dom.CompilationUnit;
+import org.eclipse.jdt.core.dom.IMethodBinding;
+import org.eclipse.jdt.core.dom.LambdaExpression;
+
+public class LambdaExpressionLocator extends ASTVisitor {
+    private CompilationUnit compilationUnit;
+    private int line;
+    private int column;
+    private boolean found;
+
+    private IMethodBinding lambdaMethodBinding;
+    private LambdaExpression lambdaExpression;
+
+    public LambdaExpressionLocator(CompilationUnit compilationUnit, int line, int column) {
+        this.compilationUnit = compilationUnit;
+        this.line = line;
+        this.column = column;
+    }
+
+    @Override
+    public boolean visit(LambdaExpression node) {
+        if (column > -1) {
+            int startPosition = node.getStartPosition();
+
+            int startColumn = this.compilationUnit.getColumnNumber(startPosition);
+            int endPosition = startPosition + node.getLength();
+            int endColumn = this.compilationUnit.getColumnNumber(endPosition);
+            int startLine = this.compilationUnit.getLineNumber(startPosition);
+            int endLine = this.compilationUnit.getLineNumber(endPosition);
+
+            // lambda on same line:
+            // list.stream().map(i -> i + 1);
+            //
+            // lambda on multiple lines:
+            // list.stream().map(user
+            // -> user.isSystem() ? new SystemUser(user) : new EndUser(user));
+
+            if ((startLine == endLine && column >= startColumn && column <= endColumn && line == startLine)
+                    || (startLine != endLine && line >= startLine && line <= endLine
+                            && (column >= startColumn || column <= endColumn))) {
```

Comment:
this condition only works for expression-style lambda with 2 lines.

Reply:
I tested for

```
        Arrays.asList(1, 2, 3).stream().map(name -> name).forEach(🔴(item) -> {
            System.out.println(item + 1 + 2 + 3 + 4 + 5);
            System.out.println(item + 1 + 2 + 3 + 4 + 5);
            System.out.println(item + 1 + 2 + 3 + 4 + 5);
            System.out.println(item + 1 + 2 + 3 + 4 + 5);
        });
```

and

```
        Arrays.asList(1, 2, 3).stream().map(name -> name).forEach(
            (item) -> 🔴{
            System.out.println(item + 1 + 2 + 3 + 4 + 5);
            System.out.println(item + 1 + 2 + 3 + 4 + 5);
            System.out.println(item + 1 + 2 + 3 + 4 + 5);
            System.out.println(item + 1 + 2 + 3 + 4 + 5);
        });
```

Its working, can you give a example snippet ?

Reply:
Case 1:
![image](https://user-images.githubusercontent.com/14052197/182273568-810a7d5d-9b93-4592-99f6-59bd022ac251.png)
This will hit lambda entry on line 21.

Case 2:
![image](https://user-images.githubusercontent.com/14052197/182274748-b41044e8-59bb-4ece-9f59-6f68b9f858d7.png)
This will hit lambda entry on line 22.

> (startLine != endLine && line >= startLine && line <= endLine
                            && (column >= startColumn || column <= endColumn))

It is not accurate to compare the column number for multi-line lambda. It might be better to compare the offset.

Reply:
True. But I couldn’t find a method to calculate offset by providing the line and column. But will check more to see if i have missed any such method. JDT should have such methods, not sure if they are part of the API.

Reply:
You can use `this.compilationUnit#getPosition(line, column)` to get the offset.

Reply:
Thanks. I think i missed it. The problem with current condition is I didn’t intend to support blocks. But still if we only do boundry check and ignore middle lines it will work when we have multi lines. But position is much simpler. I will switch to that.

Reply:
For the scenario 1 as user what you would expect ?

Reply:
> For the scenario 1 as user what you would expect ?

I expect adding a breakpoint on (Line 22, Column 24), and if Column 24 breakpoint is not supported, then fallback to a line breakpoint on Line 22. 

If it's inside a block body, I'll treat it as a normal function body and not fall back to the Lambda entry breakpoint.

Reply:
LambdaExpression.getBody() will return either a Block or an Expression, which we can use to determine which lambda it is.

If it's an expression-style lambda, it's ok to enable Lambda breakpoint if the offset is within the whole LambdaExpression range.
If it's a block-style lambda, I would enable Lambda breakpoint only if the offset is in [LambdaStart, LambdaBodyStart).

Just my two cents.

Reply:
Now keeping mind the future expansion of inline breakpoints for method invocations, I decided to restrict the lambda breakpoints by
- only support for expressions
- the inline breakpoint should be before the expression

This way we will not break any functionality nor complicate breakpoint logic in future when we introduce method invocation inline breakpoints. And for lambda blocks we can always use line breakpoints which is much clear and straight forward.

WDYT ?